gadget/install/install.go

```diff
@@ -510,6 +551,96 @@ func WriteContent(onVolumes map[string]*gadget.Volume, observer gadget.ContentOb
 	return onDiskVols, nil
 }
 
+func FinishEncryption(model gadget.Model, setupData *EncryptionSetupData) error {
+	// Save storage traits to save/data partitions
+	encryptionParams := map[string]gadget.StructureEncryptionParameters{}
+	var volName string
+	for name, p := range setupData.parts {
+		encryptionParams[name] = p.encryptionParams
+		volName = p.volName
+	}
+	optsPerVol := map[string]*gadget.DiskVolumeValidationOptions{
+		// this assumes that the encrypted partitions above are always only on the
+		// system-boot volume, this assumption may change
+		volName: {
+			ExpectedStructureEncryption: encryptionParams,
+		},
+	}
+	// save the traits to ubuntu-data host and optionally to ubuntu-save if it exists
+	// FIXME ubuntu-save forced to true
+	if err := saveStorageTraits(model, setupData.laidOutVols, optsPerVol, true); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func KeysForRole(setupData *EncryptionSetupData) map[string]keys.EncryptionKey {
+	keyForRole := make(map[string]keys.EncryptionKey)
+	for _, p := range setupData.parts {
+		keyForRole[p.onDiskStruct.Role] = p.encryptionKey
+	}
+	return keyForRole
+}
+
+func EncryptPartitions(onVolumes map[string]*gadget.Volume, gadgetRoot, kernelRoot string,
+	model *asserts.Model, encryptionType secboot.EncryptionType, perfTimings timings.Measurer) (*EncryptionSetupData, error) {
+
+	_, allLaidOutVols, err := gadget.LaidOutVolumesFromGadget(gadgetRoot, kernelRoot, model)
```

Comment:
we need TODOs in the places we call this because in a world with "partial" we will need to lay out the volumes based on the input onVolumes and not what is in the gadget, after checking though that they are compatible